### DISCUSSION TITLE
Linux 6.0 rocks

### DISCUSSION THREAD

**Discussion by uNetworkingAB** [axboe/liburing#859] [Show and tell]
These numbers are insane,

With epoll (highly optimized usage) I could formerly do TCP localhost ping/pong between two processes, small messages, on one CPU-core per process, with 325k messages/second.

With latest features of Linux 6.0 and liburing, I'm now doing 424k messages per second. This is on a 10 year old mid-range CPU.

I used to have a lot of problems getting io_uring to perform with former kernels but since 6.0 it's just completely obliterating epoll for me.

![image](https://github.com/axboe/liburing/assets/110806833/033aa4e3-a6ed-4d23-b5eb-9e75050254e3)

**Comment by da-line** [comment]
Great achievement 👍 Would be amazing if you could share us some more details/code, what io_uring features you used SQPOLL, file or buffer registration?  Did you compare system cpu/memory usage between epoll/io_uring?

**Reply by uNetworkingAB** [reply]
I'm still at POC stage but I have something that I will build on in the following weeks. These are the features I've found work best:

* io_uring_register_buf_ring
* io_uring_register_files_sparse
* io_uring_register_ring_fd
* params.flags = IORING_SETUP_COOP_TASKRUN | IORING_SETUP_SINGLE_ISSUER;
* io_uring_submit_and_wait
* io_uring_for_each_cqe
* io_uring_prep_recv_multishot
* io_uring_prep_multishot_accept_direct
* io_uring_prep_send (haven't seen anything better for my case)

I think this is all. But prep_send_zc looks interesting for large sends.
